pkg/nsx/services/nsxserviceaccount/cluster.go

```diff
@@ -0,0 +1,278 @@
+	nsxServiceAccountService.PrincipalIdentityStore = cache.NewIndexer(KeyFunc, cache.Indexers{common.TagScopeNSXServiceAccountCRUID: IndexFunc})
+	nsxServiceAccountService.ClusterControlPlaneStore = cache.NewIndexer(KeyFunc, cache.Indexers{common.TagScopeNSXServiceAccountCRUID: IndexFunc})
+
+	go queryPrincipalIdentity(nsxServiceAccountService, &wg, fatalErrors)
+	go queryClusterControlPlane(nsxServiceAccountService, &wg, fatalErrors)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		return nsxServiceAccountService, err
+	}
+
+	return nsxServiceAccountService, nil
+}
+
+func (s *NSXServiceAccountService) CreateOrUpdateNSXServiceAccount(ctx context.Context, obj *v1alpha1.NSXServiceAccount) error {
+	clusterName := s.getClusterName(obj.Namespace, obj.Name)
+	normalizedClusterName := util.NormalizeId(clusterName)
+	// TODO: Support default project name in config as project
+	project := s.NSXConfig.CoeConfig.Cluster
+	vpcName := obj.Namespace
+	vpcPath := fmt.Sprintf("/orgs/default/projects/%s/vpcs/%s", util.NormalizeId(project), vpcName)
+
+	// generate certificate
+	subject := util.DefaultSubject
+	subject.CommonName = clusterName
+	cert, key, err := util.GenerateCertificate(&subject, util.DefaultValidDays)
+	if err != nil {
+		return err
+	}
+
+	// create PI
+	if _, exists, _ := s.PrincipalIdentityStore.GetByKey(normalizedClusterName); !exists {
+		pi, err := s.NSXClient.WithCertificateClient.Create(mpmodel.PrincipalIdentityWithCertificate{
+			IsProtected: &isProtectedTrue,
+			Name:        &normalizedClusterName,
+			NodeId:      &normalizedClusterName,
+			Role:        nil,
+			RolesForPaths: []mpmodel.RolesForPath{{
+				Path: &readerPath,
+				Roles: []mpmodel.Role{{
+					Role: &readerRole,
+				}},
+			}, {
+				Path: &vpcPath,
+				Roles: []mpmodel.Role{{
+					Role: &vpcRole,
+				}},
+			}},
+			CertificatePem: &cert,
+			Tags:           common.ConvertTagsToMPTags(s.buildBasicTags(obj)),
+		})
+		if err != nil {
+			return err
+		}
+		s.PrincipalIdentityStore.Add(pi)
+	}
+
+	// create ClusterControlPlane
+	clusterId := ""
+	if _, exists, _ := s.ClusterControlPlaneStore.GetByKey(normalizedClusterName); !exists {
+		ccp, err := s.NSXClient.ClusterControlPlanesClient.Update(siteId, enforcementpointId, normalizedClusterName, model.ClusterControlPlane{
+			Revision:     &revision1,
+			ResourceType: &antreaClusterResourceType,
+			Certificate:  &cert,
+			VhcPath:      &vpcPath,
+			Tags:         s.buildBasicTags(obj),
+		})
+		if err != nil {
+			return err
+		}
+		s.ClusterControlPlaneStore.Add(ccp)
+		clusterId = *ccp.NodeId
+	}
+
+	// create Secret
+	secretName := obj.Name + SecretSuffix
+	secretNamespace := obj.Namespace
+	if err := s.Client.Create(ctx, &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: secretNamespace,
+			// TODO: Add labels/annotations
+			Labels:          nil,
+			Annotations:     nil,
+			OwnerReferences: nil,
```

Comment:
Should it set ownerReference to the corresponding NSXSA CR?

Reply:
Sure, please check if my implementation is expected.